internal/oci/oci.go

```diff
@@ -0,0 +1,135 @@
+	ociremote "github.com/sigstore/cosign/pkg/oci/remote"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	coptions "github.com/sigstore/cosign/cmd/cosign/cli/options"
+	"github.com/sigstore/cosign/pkg/cosign"
+	"github.com/sigstore/cosign/pkg/oci"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sigstore/sigstore/pkg/signature"
+)
+
+// options is a struct that holds options for verifier.
+type options struct {
+	PublicKey []byte
+	Keychain  authn.Keychain
+	Context   context.Context
+}
+
+// Options is a function that configures the options applied to a Verifier.
+type Options func(opts *options)
+
+// WithPublicKey sets the public key.
+func WithPublicKey(publicKey []byte) Options {
+	return func(opts *options) {
+		opts.PublicKey = publicKey
+	}
+}
+
+func WithAuthnKeychain(keychain authn.Keychain) Options {
+	return func(opts *options) {
+		opts.Keychain = keychain
+	}
+}
+
+func WithContext(ctx context.Context) Options {
+	return func(opts *options) {
+		opts.Context = ctx
+	}
+}
+
+// Verifier is a struct which is responsible for executing verification logic.
+type Verifier struct {
+	opts    *cosign.CheckOpts
+	context context.Context
+}
+
+// New initializes a new Verifier.
+func New(opts ...Options) (*Verifier, error) {
+	o := options{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	checkOpts := &cosign.CheckOpts{}
+
+	ro := coptions.RegistryOptions{}
+	co, err := ro.ClientOpts(o.Context)
+	if err != nil {
+		return nil, err
+	}
+
+	if o.Keychain != nil {
+		co = append(co, ociremote.WithRemoteOptions(remote.WithAuthFromKeychain(o.Keychain)))
+	}
+
+	checkOpts.RegistryClientOpts = co
+
+	// If a public key is provided, it will use it to verify the signature.
+	// If there is no public key provided, it will try keyless verification.
+	// https://github.com/sigstore/cosign/blob/main/KEYLESS.md.
+	if len(o.PublicKey) > 0 {
+		pubKeyRaw, err := cryptoutils.UnmarshalPEMToPublicKey(o.PublicKey)
+		if err != nil {
+			return nil, err
+		}
+
+		checkOpts.SigVerifier, err = signature.LoadVerifier(pubKeyRaw, crypto.SHA256)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		rcerts, err := fulcio.GetRoots()
+		if err != nil {
+			return nil, fmt.Errorf("unable to get Fulcio root certs: %w", err)
+		}
+		checkOpts.RootCerts = rcerts
+
+		icerts, err := fulcio.GetIntermediates()
+		if err != nil {
+			return nil, fmt.Errorf("unable to get Fulcio intermediate certs: %w", err)
+		}
+		checkOpts.IntermediateCerts = icerts
+
+		rc, err := rekor.NewClient(coptions.DefaultRekorURL)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create Rekor client: %w", err)
+		}
+		checkOpts.RekorClient = rc
+	}
+
+	return &Verifier{
+		opts:    checkOpts,
+		context: o.Context,
+	}, nil
+}
+
+// VerifyImageSignatures verify the authenticity of the given ref OCI image.
+func (v *Verifier) VerifyImageSignatures(ref name.Reference) ([]oci.Signature, bool, error) {
+	return cosign.VerifyImageSignatures(v.context, ref, v.opts)
```

Comment:
Think it would be better to do:

```suggestion
func (v *Verifier) VerifyImageSignatures(ctx context.Context, ref name.Reference) ([]oci.Signature, bool, error) {
	return cosign.VerifyImageSignatures(ctx, ref, v.opts)
```

If the caller wants to re-use their context (with e.g. a timeout), they can simply pass in the same context. This also seems to prevent the context from ever being `nil`, which appears to be an option now when `WithContext` is not passed.